aten/src/ATen/native_parse.py

```diff
@@ -88,11 +98,35 @@ def parse_arguments(args, func_decl, func_name, func_return):
         # be better if we just named everything and matched by name.
         if is_out_fn and arg_idx < len(func_return):
             argument_dict['output'] = True
+        argument_dict["type"], argument_dict["annotation"] = get_type_annotation(argument_dict["type"])
+        # We assume that type annotations are only used for output tensors
+        is_output_arg = False
+        for ret in func_return:
+            if ret["annotation"] and ret["annotation"] == argument_dict["annotation"]:
+                argument_dict["output"] = True
+                output_arguments.append(argument_dict)
+                assert not is_output_arg
+                is_output_arg = True
+                break
+        if is_output_arg:
+            is_out_fn = True
+            continue
         if kwarg_only:
             argument_dict['kwarg_only'] = True
 
         arguments.append(argument_dict)
-    return arguments
+
+    arguments = output_arguments + arguments
+
+    if is_out_fn and func_decl.get('variants', []) not in [[], 'function', ['function']]:
+        raise RuntimeError("Native functions suffixed with _out MUST be declared with only the function variant; "
+                           "e.g., variants: function; otherwise you will tickle a Python argument binding bug "
+                           "(which usually manifests itself as the result variable being undefined.) "
+                           "The culprit was: {}".format(func_name))
+    if is_out_fn:
+        func_name += "_out"
```

Comment:
I think you should just move this outside.  You can either pass that this is an out function or just check explicitly later (there are output arguments and this isn't an inplace function?).